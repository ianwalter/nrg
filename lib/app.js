// Load environment variables from .env file.
require('dotenv').config()

const util = require('util')
// const fs = require('fs')
// const path = require('path')
const http = require('http')
const https = require('https')
const merge = require('@ianwalter/merge')
const Koa = require('koa')
const json = require('koa-json')
const bodyParser = require('koa-bodyparser')
const compress = require('koa-compress')
const pino = require('pino')
const pinoMiddleware = require('koa-pino-logger')
const { addRouter } = require('@ianwalter/nrg-router')
const knex = require('knex')
const { Model, knexSnakeCaseMappers } = require('objection')
const { SchemaValidator } = require('@ianwalter/correct')
// const ejs = require('ejs')
const supertest = require('supertest')
const { Print } = require('@ianwalter/print')
const Url = require('@ianwalter/url')
const { oneLine } = require('common-tags')
const bcrypt = require('bcrypt')

const Account = require('./models/Account')
const Token = require('./models/Token')
const { serveStatic, serveWebpack } = require('./middleware/client')
const { handleError } = require('./middleware/error')

// If running in a CLI context, parse the JSON string to get the CLI options.
const nrgCli = process.env.NRG_CLI && JSON.parse(process.env.NRG_CLI)

// Determine the "mode" the application is running in.
const isNotDev = process.env.NODE_ENV && process.env.NODE_ENV !== 'development'
const isTest = process.env.NODE_ENV === 'test'

const defaults = {
  env: process.env.NRG_ENV || process.env.APP_ENV,
  isNotDev,
  isTest,
  log: {
    level: isNotDev ? (isTest ? 'error' : 'info') : 'debug',
    unhandled: true,
    prettyPrint: isNotDev && !isTest ? false : { stream: false },
    prettifier: isNotDev && !isTest ? false : require('@ianwalter/pino-print'),
    redact: isNotDev ? ['req.headers.cookie', 'res.headers["set-cookie"]'] : []
  },
  baseUrl: isNotDev ? 'http://0.0.0.0:9999' : 'http://localhost:9999',
  handleError,
  router: true,
  health: { path: '/health' },
  prettyJson: !isNotDev,
  compress: true,
  static: {
    enabled: false,
    options: {},
    webpack: {
      enabled: false,
      options: {
        devMiddleware: {
          serverSideRender: true,
          publicPath: '/static'
        }
      }
    }
  },
  sessions: { keys: false },
  db: { client: 'pg', ...knexSnakeCaseMappers() },
  hash: { bytes: 48, rounds: 12 },
  accounts: {
    enabled: false,
    dummyPassword: 'ijFu54r6PyNdrNLj9yoBu',
    models: {
      Account,
      Token
    },
    emails: {}
  }
}

module.exports = function createApp (options = {}) {
  // Combine app-supplied and CLI-supplied options.
  merge(options, nrgCli)

  // Combine defaults and user-supplied options.
  options = merge({}, defaults, options)

  // Create the Koa app instance.
  const app = new Koa()

  // Make it easy to determine if the app is running in "production" or
  // "development" mode.
  app.isNotDev = app.context.isNotDev = options.isNotDev

  // TODO: comment
  app.silent = true

  // Add the options to the app context so that they can be referenced elsewhere
  // in the app.
  app.context.options = options

  // Determine which logger to use depending on the passed configuration and/or
  // runtime environment.
  const usePino = options.log && !nrgCli && !options.logger
  app.logger = usePino
    ? pino(options.log)
    : options.logger || new Print(options.log || defaults.log)

  if (usePino) {
    // Use thee koa pino middleware.
    app.use(pinoMiddleware({ logger: app.logger }))
  } else if (!nrgCli) {
    // If a custom logger is being used, add it to the context.
    app.use((ctx, next) => (ctx.log = app.logger) && next())
  }

  // If configured, set up handlers for any uncaught exceptions and unhandled
  // Promise rejections that happen within the current process.
  if (options.log && options.log.unhandled && usePino) {
    const syncLogger = pino(options.log, pino.destination(2))
    const finalLogger = pino.final(syncLogger, (err, log) => log.error(err))
    process.on('unhandledRejection', finalLogger)
    process.on('uncaughtException', finalLogger)
  } else if (options.log && options.log.unhandled) {
    process.on('unhandledRejection', app.logger.error)
    process.on('uncaughtException', app.logger.error)
  }

  app.logger.debug('Options', util.inspect(options, false, 10, true))

  // Add error-handling middleware.
  if (!nrgCli) {
    app.use(async function errorHandlingMiddleware (ctx, next) {
      try {
        await next()
      } catch (err) {
        // Pass the error and context to the configured error handler.
        await options.handleError(err, ctx)
      }
    })
  }

  // TODO:
  app.context.baseUrl = new Url(options.baseUrl)

  // TODO: comment
  if (!nrgCli && options.static.enabled) {
    if (app.isNotDev || !options.static.webpack.enabled) {
      //
      app.use(serveStatic(options.static.options))
    } else {
      //
      app.use(serveWebpack(options.static.webpack.options))
    }
  }

  // Add a knex database instance to the server context and tell Objection to
  // use that instance.
  if (typeof options.db === 'string' || options.db.connection) {
    if (typeof options.db.migrations === 'string') {
      options.db.migrations = { directory: options.db.migrations }
    }
    if (typeof options.db.seeds === 'string') {
      options.db.seeds = { directory: options.db.seeds }
    }

    app.db = app.context.db = knex(options.db)
    Model.knex(app.db)
  }

  // TODO
  if (options.accounts.enabled) {
    const { keys } = options.sessions
    if (!keys || !keys.length || !Array.isArray(keys) || !keys[0]) {
      app.logger.fatal(oneLine`
        In order to use accounts, you have to configure sessions.keys to an
        array of secret strings used to encrypt session tokens.
      `)
      process.exit(1)
    }

    const salt = bcrypt.genSaltSync(options.hash.rounds)
    app.context.options.accounts.dummyPassword = bcrypt.hashSync(
      options.accounts.dummyPassword,
      salt
    )

    const { Account } = options.accounts.models
    app.context.validators = {
      login: new SchemaValidator(Account.loginSchema),
      registration: new SchemaValidator(Account.registrationSchema)
    }

    // If not already set, create render functions for the email verification
    // and password reset emails by compiling the EJS templates.
    // const emailsPath = path.join(__dirname, '..', 'emails', 'dist')
    // if (!options.accounts.emails.emailVerification) {
    //   options.accounts.emails.emailVerification = ejs.compile(
    //     fs.readFileSync(
    //       path.join(emailsPath, 'emailVerification.ejs'),
    //       'utf8'
    //     )
    //   )
    // }
    // if (!options.accounts.emails.passwordReset) {
    //   options.accounts.emails.passwordReset = ejs.compile(
    //     fs.readFileSync(path.join(emailsPath, 'passwordReset.ejs'), 'utf8')
    //   )
    // }

    // Add a renderEmail function to the context that can be used to call the
    // render email render functions from the middleware.
    app.context.renderEmail = function renderEmail (key, data) {
      return options.accounts.emails[key](data)
    }
  }

  if (!nrgCli) {
    // Add the bodyparser middleware that can parse a request body into json,
    // etc.
    app.use(bodyParser({ enableTypes: ['json', 'form', 'text'] }))

    // Use middleware that automatically pretty-prints JSON responses when not
    // in production.
    if (options.prettyJson) {
      app.use(json({ pretty: true }))
    }

    // Return compressed responses for clients that identify as being able to
    // handle them.
    if (options.compress) {
      app.use(compress(options.compress))
    }

    // TODO
    if (options.sessions.keys) {
      const redisStore = require('koa-redis')
      const session = require('koa-session')
      const CSRF = require('koa-csrf')
      const store = redisStore({ url: options.sessions.redisUrl })
      app.keys = options.sessions.keys
      app.use(session({ store }, app))
      app.use(new CSRF())
    }
  }

  // Add routing to the app via nrg-router.
  if (options.router) {
    addRouter(app)

    // Add a simple health check route.
    if (options.health) {
      app.get(options.health.path, (ctx, next) => (ctx.status = 200))
    }
  }

  // Add a start method to the app that makes it easy to start listening for
  // connections.
  app.start = function start () {
    const protocol = app.context.baseUrl.protocol === 'https:' ? https : http
    const server = protocol.createServer(app.callback())
    server.listen(app.context.baseUrl.port, app.context.baseUrl.hostname)
    app.logger.info('nrg now serving requests at:', app.context.baseUrl.href)
    app.server = server
  }

  app.test = function test (url, options = {}) {
    const request = supertest(this.callback())
    return {
      post (data) {
        return this.run(request.post(url).set(options).send(data))
      },
      put (data) {
        return this.run(request.put(url).set(options).send(data))
      },
      get () {
        return this.run(request.get(url).set(options))
      },
      delete () {
        return this.run(request.delete(url).set(options))
      },
      run (request) {
        return new Promise(resolve => request.end((_, res) => resolve(res)))
      }
    }
  }

  // Return the app instance.
  return app
}
