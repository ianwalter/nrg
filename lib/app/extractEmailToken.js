const clone = require('@ianwalter/clone')
const cheerio = require('cheerio')
const replaceAll = require('replace-string')

async function extractEmailToken (byEmail, selector = '.button') {
  const email = clone(await this.getTestEmail(byEmail))

  const $ = cheerio.load(email.html)
  const url = new URL($(selector).attr('href'))
  const token = url.searchParams.get('token')

  // Replace the token string with an arbitrary string so that the email body
  // can more easily be used with a snapshot matcher.
  email.html = replaceAll(email.html, token, '<~ TOKEN WAS HERE ~>')

  return { email, token }
}

module.exports = extractEmailToken
