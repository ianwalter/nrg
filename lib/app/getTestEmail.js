const { requester } = require('@ianwalter/requester')

module.exports = async function getTestEmail (byEmail) {
  const { host = 'localhost', port = 1080 } = this.cfg.email.transport
  const { body } = await requester.get(`http://${host}:${port}/email`)
  const email = body.find(byEmail)
  if (email) await requester.delete(`http://${host}:${port}/email/${email.id}`)
  return email
}
