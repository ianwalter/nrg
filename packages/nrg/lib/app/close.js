export function install (app, ctx) {
  ctx.logger.debug('TODO:')

  app.close = function close () {
    // Close any open database connections.
    if (this.db) this.db.destroy()

    // Close any open redis connections.
    if (this.redis) this.redis.quit()

    if (this.mq) {
      // Silence channel ended error.
      this.mq.channel.on('error', err => this.logger.ns('nrg.mq').debug(err))

      // Close message queue connection.
      this.mq.connection.close()
    }
  }
}
