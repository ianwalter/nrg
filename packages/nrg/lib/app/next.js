module.exports = async function next (req, res, getServerSideProps) {
  req.getServerSideProps = getServerSideProps
  await this.callback()(req, res)
  return res.serverSideProps || res
}
