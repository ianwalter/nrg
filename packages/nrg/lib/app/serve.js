const http = require('http')
const getHostUrl = require('../utilities/getHostUrl')

module.exports = function serve (port, hostname) {
  // Create the server instance by specifying the app's callback as the handler.
  const server = http.createServer(this.callback())

  // Prefer the port and hostname passed as arguments to those configured in the
  // app even if the port is 0 (which means use a random unused port) so that
  // they can be overridden when serving (e.g. in a test).
  const portToUse = port !== undefined ? port : (this.context.cfg.port || 0)
  const hostnameToUse = hostname || this.context.cfg.hostname

  return new Promise(resolve => {
    server.listen(portToUse, hostnameToUse, err => {
      if (err) {
        if (this.log) this.log.error(err)
        process.exit(1)
      }

      // Set the server URL (the local URL which can be different from the
      // base URL) so that whatever is starting the server (e.g. tests) can
      // easily know what URL to use.
      server.url = getHostUrl(hostnameToUse, portToUse || server.address().port)

      if (this.log) {
        this.log
          .ns('nrg.server')
          .info(`${this.context.cfg.name} server started:`, server.url)
      }

      // Add a destroy method to the server instance if not in a production
      // environment (e.g. development or test).
      // https://github.com/nodejs/node/issues/2642
      // Logic adapted from: https://github.com/isaacs/server-destroy
      if (!this.context.cfg.isProd) {
        const sockets = []

        // Keep track of all active connections.
        server.on('connection', socket => {
          const index = sockets.push(socket) - 1
          socket.on('close', () => sockets.splice(index, 1))
        })

        // Add a destroy method to the server instance that closes the server
        // and destroys all active connections.
        server.destroy = () => new Promise(resolve => server.close(() => {
          for (const socket of sockets) socket.destroy()
          setTimeout(resolve)
        }))
      }

      // Return the server instance.
      resolve(server)
    })
  })
}
