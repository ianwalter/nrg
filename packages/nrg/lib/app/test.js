module.exports = function test (url, options = {}) {
  const app = this
  const supertest = require('supertest')
  const cb = app.callback()
  const request = supertest(cb)

  // If options is a re-used response object, re-use cookie and CSRF token
  // values in request headers.
  if (options.status) {
    // Supertest does not like a undefined headers.
    const headers = options.request.header
    const Cookie = options.headers['set-cookie'] || headers.Cookie
    const csrf = headers['csrf-token']
    options = {
      ...Cookie ? { Cookie } : {},
      ...csrf ? { 'csrf-token': csrf } : {}
    }
    if (app.log) app.log.ns('nrg.test').debug('Test request options', options)
  }

  return {
    post (data) {
      return this.runWithCsrf(request.post(url).set(options).send(data))
    },
    put (data) {
      return this.runWithCsrf(request.put(url).set(options).send(data))
    },
    get () {
      return this.run(request.get(url).set(options))
    },
    delete () {
      return this.runWithCsrf(request.delete(url).set(options))
    },
    run (request) {
      return new Promise(resolve => request.end((_, res) => resolve(res)))
    },
    async runWithCsrf (request) {
      if (app.keys?.length && !options['csrf-token']) {
        if (app.log) {
          app.log.ns('nrg.test').debug('Adding CSRF token for test')
        }
        const response = await supertest(cb).get('/csrf-token').set(options)
        request.set('Cookie', response.headers['set-cookie'])
        request.set('csrf-token', response.body.csrfToken)
      }
      return this.run(request)
    }
  }
}
