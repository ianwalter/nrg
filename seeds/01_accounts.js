const bcrypt = require('bcrypt')

const password = 'iJustC4n7!gnore'
const salt = bcrypt.genSaltSync(12)
const encryptedPassword = bcrypt.hashSync(password, salt)
const accounts = [
  {
    id: 1,
    firstName: 'General',
    lastName: 'Test',
    email: 'general_test@example.com',
    password: encryptedPassword,
    emailVerified: true
  },
  {
    id: 2,
    firstName: 'Password Reset',
    lastName: 'Test',
    email: 'password_reset_test@example.com',
    password: encryptedPassword,
    emailVerified: true
  },
  {
    id: 3,
    firstName: 'Account Update',
    lastName: 'Test',
    email: 'account_update_test@example.com',
    password: encryptedPassword,
    emailVerified: true
  },
  {
    id: 4,
    firstName: 'Unverified',
    lastName: 'Test',
    email: 'unverified_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  },
  {
    id: 5,
    firstName: 'Disabled',
    lastName: 'Test',
    email: 'disabled_test@example.com',
    password: encryptedPassword,
    emailVerified: true,
    enabled: false
  },
  {
    id: 6,
    firstName: 'Owner',
    lastName: 'Test',
    email: 'owner_test@example.com',
    password: encryptedPassword,
    emailVerified: true,
    enabled: true
  },
  {
    id: 7,
    firstName: 'Admin',
    lastName: 'Test',
    email: 'admin_test@example.com',
    password: encryptedPassword,
    emailVerified: true,
    enabled: true
  },
  {
    id: 8,
    firstName: 'Change Password',
    lastName: 'Test',
    email: 'change_password_test@example.com',
    password: encryptedPassword,
    emailVerified: true,
    enabled: true
  },
  {
    id: 9,
    firstName: 'Read Only',
    lastName: 'Test',
    email: 'read_only_test@example.com',
    password: encryptedPassword,
    emailVerified: true,
    enabled: true
  },
  {
    id: 10,
    firstName: 'Change Email',
    lastName: 'Test',
    email: 'change_email_test@example.com',
    password: encryptedPassword,
    emailVerified: true,
    enabled: true
  },
  {
    id: 11,
    firstName: 'Previous Email',
    lastName: 'Token Test',
    email: 'previous_email_token_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  },
  {
    id: 12,
    firstName: 'Expired Email',
    lastName: 'Token Test',
    email: 'expired_email_token_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  },
  {
    id: 13,
    firstName: 'Wrong Email',
    lastName: 'Token Test',
    email: 'wrong_email_token_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  },
  {
    id: 14,
    firstName: 'Mismatch Email',
    lastName: 'Token Test',
    email: 'mismatch_email_token_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  },
  {
    id: 15,
    firstName: 'Will Verify',
    lastName: 'Test',
    email: 'will_verify_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  },
  {
    id: 16,
    firstName: 'Reset Verify',
    lastName: 'Test',
    email: 'reset_verify_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  },
  {
    id: 17,
    firstName: 'Existing Verified',
    lastName: 'Test',
    email: 'existing_verified_test@example.com',
    password: encryptedPassword,
    emailVerified: true
  },
  {
    id: 18,
    firstName: 'Existing Unverified',
    lastName: 'Test',
    email: 'existing_unverified_test@example.com',
    password: encryptedPassword,
    emailVerified: false
  }
]

module.exports = {
  password,
  accounts,
  seed: async knex => {
    await knex.raw('TRUNCATE TABLE accounts RESTART IDENTITY CASCADE')
    await knex('accounts').insert(accounts)
    await knex.raw(`
      SELECT setval('accounts_id_seq',  (SELECT MAX(id) + 1 FROM accounts))
    `)
  }
}
