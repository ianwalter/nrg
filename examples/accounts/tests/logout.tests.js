const { test } = require('@ianwalter/bff')
const app = require('../app')
const { accounts, password } = require('../seeds/01_accounts')

const generalUser = accounts.find(a => a.firstName === 'General')

test('Logout when not logged in', async t => {
  const response = await app.test('/logout').delete()
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body.csrfToken.length).toBeGreaterThan(0)
})

test('Logout when logged in', async t => {
  // Login.
  let response = await app.test('/login').post({ ...generalUser, password })

  // Verify account data can be retrieved.
  response = await app.test('/account', response).get()
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body).toMatchSnapshot()

  // Logout.
  response = await app.test('/logout', response).delete()
  const { csrfToken } = response.body
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body.csrfToken.length).toBeGreaterThan(0)

  // Verify account data cannot be retrieved.
  response = await app.test('/account', response).get()
  t.expect(response.statusCode).toBe(401)

  // Verify a user can login with the returned CSRF token.
  response.request.options.headers['csrf-token'] = csrfToken
  const credentials = { ...generalUser, password }
  response = await app.test('/login', response).post(credentials)
  t.expect(response.statusCode).toBe(201)

  // Verify account data can be retrieved.
  response = await app.test('/account', response).get()
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body).toMatchSnapshot()
})
