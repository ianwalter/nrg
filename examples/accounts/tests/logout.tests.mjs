import { test } from '@ianwalter/bff'
import app from '../app/index.js'
import { accounts, password } from '../seeds/01_accounts.js'

const generalUser = accounts.find(a => a.firstName === 'General')

test('Logout • When not logged in', async t => {
  const response = await app.test('/logout').delete()
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body.csrfToken.length).toBeGreaterThan(0)
})

test('Logout • When logged in', async t => {
  // Login.
  let response = await app.test('/login').post({ ...generalUser, password })

  // Verify account data can be retrieved.
  response = await app.test('/account', response).get()
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body).toMatchSnapshot()

  // Logout.
  response = await app.test('/logout', response).delete()
  const { csrfToken } = response.body
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body.csrfToken.length).toBeGreaterThan(0)

  // Verify account data cannot be retrieved.
  response = await app.test('/account', response).get()
  t.expect(response.statusCode).toBe(401)

  // Verify a user can login with the returned CSRF token.
  response.request.options.headers['csrf-token'] = csrfToken
  const credentials = { ...generalUser, password }
  response = await app.test('/login', response).post(credentials)
  t.expect(response.statusCode).toBe(201)

  // Verify account data can be retrieved.
  response = await app.test('/account', response).get()
  t.expect(response.statusCode).toBe(200)
  t.expect(response.body).toMatchSnapshot()
})
