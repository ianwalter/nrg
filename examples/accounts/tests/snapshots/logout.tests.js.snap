// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logout • When logged in 1`] = `
Object {
  "email": "general_test@example.com",
  "emailVerified": true,
  "firstName": "General",
  "id": 1,
  "lastName": "Test",
}
`;

exports[`Logout • When logged in 2`] = `
Object {
  "email": "general_test@example.com",
  "emailVerified": true,
  "firstName": "General",
  "id": 1,
  "lastName": "Test",
}
`;
